go-controller/pkg/config/utils.go

```diff
@@ -23,30 +23,63 @@ func (e CIDRNetworkEntry) HostBits() uint32 {
 // addresses from.
 func ParseClusterSubnetEntries(clusterSubnetCmd string) ([]CIDRNetworkEntry, error) {
 	var parsedClusterList []CIDRNetworkEntry
-
+	var maxSubnetLength uint32
 	clusterEntriesList := strings.Split(clusterSubnetCmd, ",")
 
 	for _, clusterEntry := range clusterEntriesList {
 		var parsedClusterEntry CIDRNetworkEntry
 
 		splitClusterEntry := strings.Split(clusterEntry, "/")
+
+		if len(splitClusterEntry) < 2 || len(splitClusterEntry) > 3 {
+			return nil, fmt.Errorf("CIDR %q not properly formatted", clusterEntry)
+		}
+
+		var err error
+		_, parsedClusterEntry.CIDR, err = net.ParseCIDR(fmt.Sprintf("%s/%s", splitClusterEntry[0], splitClusterEntry[1]))
+		if err != nil {
+			return nil, err
+		}
+
+		if parsedClusterEntry.CIDR.IP.To4() != nil {
+			maxSubnetLength = 32
+		} else {
+			maxSubnetLength = 128
+		}
+		entryMaskLength, _ := parsedClusterEntry.CIDR.Mask.Size()
 		if len(splitClusterEntry) == 3 {
 			tmp, err := strconv.ParseUint(splitClusterEntry[2], 10, 32)
 			if err != nil {
 				return nil, err
 			}
 			parsedClusterEntry.HostSubnetLength = uint32(tmp)
-		} else if len(splitClusterEntry) == 2 {
-			// the old hardcoded value for backwards compatability
-			parsedClusterEntry.HostSubnetLength = 24
+			if parsedClusterEntry.HostSubnetLength >= maxSubnetLength {
+				return nil, fmt.Errorf("cannot use maximum subnet mask length of %d for creating host subnets", maxSubnetLength)
+			} else if parsedClusterEntry.HostSubnetLength == uint32(entryMaskLength) {
+				return nil, fmt.Errorf("cannot use the same host subnet length as the cluster subnet: %d", entryMaskLength)
+			} else if parsedClusterEntry.HostSubnetLength < uint32(entryMaskLength) {
+				return nil, fmt.Errorf("cannot use a host subnet length mask shorter than the cluster subnet mask. "+
+					"host subnet length: %d, cluster subnet length: %d", parsedClusterEntry.HostSubnetLength, entryMaskLength)
+			}
 		} else {
-			return nil, fmt.Errorf("CIDR %q not properly formatted", clusterEntry)
-		}
+			// host subnet prefix length not specified so we determine it automatically
+			// giving 3 bits to subnetting gives 8 per-host subnets which is a reasonable default
+			// we need at least 2 bits for host addresses, which means we ideally want at least 5 bits total
```

Comment:
@trozet what would say `10.0.0.0/16` get as a default hostsubnet length? If it doesn't get /24, could we find a way to make the largest default hostsubnet be 24?

Reply:
It would get a /19, giving 8 subnets. Sure I can tweak it to attempt to get more bits for subnetting, and we can make it a larger value for ipv6. The length of the mask is dependent on the cluster CIDR that we are given to achieve a desired number of subnets, so I don't think choosing /24 is the right approach. Just let me know how many subnets you think by default an ipv4 deployment should get (how many nodes), and then the code can figure out what host length prefix that ends up being, given a cluster CIDR. Also, same thing for ipv6, whats a valid default #of nodes/subnets?